setup.py

```diff
@@ -1231,43 +1234,12 @@ def run(self):
         )
         extensions.append(ext)
 
-if pypy:
-    # add dummy extension, to ensure build_ext runs
-    dummy_ext = Extension('dummy', sources=[])
-    extensions = [dummy_ext]
-
-    bld_ext = cmdclass['build_ext']
-    class pypy_build_ext(bld_ext):
-        """hack to build pypy extension only after building bundled libzmq
-
-        otherwise it will fail when libzmq is bundled.
-        """
-        def build_extensions(self):
-            self.extensions.remove(dummy_ext)
-            bld_ext.build_extensions(self)
-            # build ffi extension after bundled libzmq,
-            # because it may depend on linking it
-            if self.inplace:
-                sys.path.insert(0, '')
-            else:
-                sys.path.insert(0, self.build_lib)
-            try:
-                from zmq.backend.cffi import ffi
-            except ImportError as e:
-                warn("Couldn't get CFFI extension: %s" % e)
-            else:
-                ext = ffi.verifier.get_extension()
-                if not ext.name.startswith('zmq.'):
-                    ext.name = 'zmq.backend.cffi.' + ext.name
-                self.extensions.append(ext)
-                self.build_extension(ext)
-            finally:
-                sys.path.pop(0)
-
-
-    # How many build_ext subclasses is this? 5? Gross.
-    cmdclass['build_ext'] = pypy_build_ext
-
+if cffi:
+    cffi_modules = ['buildutils/build_cffi.py:ffi']
+elif is_pypy:
+    RuntimeError('PyPy requires cffi')
+else:
+    cffi_modules = []
```

Comment:
setuptools.setup supports a `cffi_modules` directive, which removes all the code above